3scale/APIcast
# [THREESCALE-10278] upgrade lua-resty-http to 0.17.1

### What
Fix https://issues.redhat.com/browse/THREESCALE-10278

lua-resty-http 0.17.1 is also required for https://issues.redhat.com/browse/THREESCALE-5105

This PR is mainly a refactoring of existing code so no additional integration tests/unittests are added.

### Verification Steps

1. Connect via proxy
* Build docker image from this git branch
```
make runtime-image IMAGE_NAME=apicast-test
```
* Run proxy dev environment
```
cd dev-environments/https-proxy-upstream-tlsv1.3
make certs
make gateway IMAGE_NAME=apicast-test
```

* Send request to APIcast
```
curl --resolve get.example.com:8080:127.0.0.1 -v "http://get.example.com:8080/?user_key=123"
```

Request should return 200
<details>

```
* Added get.example.com:8080:127.0.0.1 to DNS cache
* Hostname get.example.com was found in DNS cache
*   Trying 127.0.0.1...
* TCP_NODELAY set
* Connected to get.example.com (127.0.0.1) port 8080 (#0)
> GET /?user_key=123 HTTP/1.1
> Host: get.example.com:8080
> User-Agent: curl/7.61.1
> Accept: */*
> 
< HTTP/1.1 200 OK
< Content-Type: application/json
< Transfer-Encoding: chunked
< Connection: keep-alive
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Credentials: true
< Date: Tue, 30 Jan 2024 04:53:30 GMT
< Server: gunicorn/19.9.0
< 
{
  "args": {
    "user_key": "123"
  }, 
  "headers": {
    "Accept": "*/*", 
    "Host": "example.com", 
    "User-Agent": "curl/7.61.1"
  }, 
  "origin": "172.21.0.4", 
  "url": "http://example.com/get?user_key=123"
}
* Connection #0 to host get.example.com left intact
```

</details>

* Proxy should receive a CONNECT method with upstream hostname not IP:PORT

```
docker compose -p https-proxy-upstream-tlsv13 logs -f proxy
```

```
proxy-1  | CONNECT   Jan 30 04:55:57.831 [1]: Connect (file descriptor 4): 172.21.0.5
proxy-1  | CONNECT   Jan 30 04:55:57.831 [1]: Request (file descriptor 4): CONNECT example.com:443 HTTP/1.1
```

2. Fetching configuration file from 3scale
* Start dev environment

```
make development
make dependencies
```

* Run APIcast locally
```
THREESCALE_DEPLOYMENT_ENV=staging APICAST_LOG_LEVEL=debug APICAST_WORKER=1 APICAST_CONFIGURATION_LOADER=lazy APICAST_CONFIGURATION_CACHE=0 THREESCALE_PORTAL_ENDPOINT=https://token@3scale-admin.example.com ./bin/apicast
```

* Run query
```
# capture apicast IP
APICAST_IP=$(docker inspect apicast_build_0-development-1 | yq e -P '.[0].NetworkSettings.Networks.apicast_build_0_default.IPAddress' -)

curl -i -k -H "Host: default-product.staging.example.com:443" "http://${APICAST_IP}:8080/?user_key=<user_key>"
```

Replace `<user_key>` with the actual user key.  The response should be HTTP/1.1 200 OK

```
HTTP/1.1 200 OK                    
Server: openresty                  
Date: Tue, 30 Jan 2024 05:08:26 GMT
Content-Type: application/json     
Content-Length: 702                
Connection: keep-alive             
x-3scale-echo-api: echo-api/1.0.3  
vary: Origin                       
x-content-type-options: nosniff    
x-envoy-upstream-service-time: 0   
```

From the APIcast log, the request to fetch configuration file return 200

```
2024/02/14 01:42:12 [debug] 763706#763706: *2 remote_v2.lua:268: proxy_configs_per_page(): proxy configs get status: 200 url: https://user_key@3scale-admin.example.con/admin/api/account/proxy_configs/production.json?host=default-product.staging.example.com&page=1&per_page=500&version=latest body: {"proxy_configs":  .... }
```
<details>

```
{
    "services": [
        {
            "proxy": {
                "error_headers_auth_failed": "text\/plain; charset=us-ascii",
                "error_headers_limits_exceeded": "text\/plain; charset=us-ascii",
                "error_headers_auth_missing": "text\/plain; charset=us-ascii",
                "error_headers_no_match": "text\/plain; charset=us-ascii",
                "error_status_no_match": 404,
                "error_status_auth_failed": 403,
                "error_status_limits_exceeded": 429,
                "error_status_auth_missing": 403,
                "secret_token": "Shared_secret_sent_from_proxy_to_API_backend_09df7e84d9ba36d8",
                "hostname_rewrite": null,
                "oidc_issuer_endpoint": null,
                "jwt_claim_with_client_id": null,
                "jwt_claim_with_client_id_type": null,
                "auth_user_key": "user_key",
                "auth_app_id": "app_id",
                "auth_app_key": "app_key",
                "oauth_login_url": null,
                "proxy_rules": [
                    {
                        "http_method": "GET",
                        "pattern": "\/",
                        "delta": 1,
                        "redirect_url": null,
                        "querystring_parameters": {},
                        "position": 1,
                        "parameters": {},
                        "metric_system_name": "hits",
                        "last": false,
                        "owner_type": "Proxy"
                        ...
                    }
                ],
                "error_auth_missing": "Authentication parameters missing",
                "api_test_path": "\/",
                "api_test_success": null,
                "apicast_configuration_driven": true,
                "oidc_issuer_type": "keycloak",
                "staging_domain": "default-product.staging.example.com",
                "production_domain": "default-product.production.example.com",
                "endpoint": "https:\/\/default-product.production.example.com: 443",
                "error_limits_exceeded": "Usage limit exceeded",
                "deployed_at": null,
                "backend": {
                    "endpoint": "https:\/\/su1.3scale.net",
                    "host": "su1.3scale.net"
                },
                "error_no_match": "No Mapping Rule matched",
                "valid?": true,
                "service_backend_version": "1",
                "hosts": [
                    "default-product.production.example.com",
                    "default-product.staging.example.com"
                ],
                "error_auth_failed": "Authentication failed",
                "lock_version": 1,
                "policy_chain": [
                    {
                        "name": "apicast",
                        "configuration": {},
                        "version": "builtin"
                    }
                ],
                "endpoint_port": 443,
                "sandbox_endpoint": "https:\/\/default-product.staging.example.com: 443",
                "authentication_method": "1",
                "api_backend": "https:\/\/echo-api.3scale.net: 443",
                "credentials_location": "query",
                "hostname_rewrite_for_sandbox": "echo-api.3scale.net"
            },          
            "backend_authentication_type": "service_token",
            "description": "",
            "name": "Default Product",
            "backend_version": "1",
            "proxiable?": true,
            "system_name": "default_product",
            ...
        },
    ]
}
```
</details>
